src/coreclr/jit/optimizer.cpp

```diff
@@ -7608,12 +7608,791 @@ bool Compiler::optIsRangeCheckRemovable(GenTree* tree)
+        ehUpdateForDeletedBlock(b3);
+    }
+
+    if (!optReturnBlock)
+    {
+        // Update bbRefs and bbPreds
+        //
+        // Replace pred 'b2' for 'b2->bbNext' with 'b1'
+        // Remove  pred 'b2' for 'b2->bbJumpDest'
+        fgReplacePred(b2->bbNext, b2, b1);
+        fgRemoveRefPred(b2->bbJumpDest, b2);
+    }
+
+    // Update loop table
+    fgUpdateLoopsAfterCompacting(b1, b2);
+    if (optReturnBlock)
+    {
+        fgUpdateLoopsAfterCompacting(b1, b3);
+    }
+}
+
+//
+//  optOptimizeBoolsReturnBlock: Optimize boolean when b1.bbJumpKind is BBJ_COND and b2.bbJumpKind is BBJ_RETURN
+//
+//  Arguments:
+//      pOptBoolsDsc    The descriptor for boolean optimization with b1, b2 and b3 members set when called
+//
+//  Returns:
+//      changeReturn    Set to true if boolean optimization is done and b1, b2 and b3 are folded into b1.
+//
+//  Notes:
+//      Below members of pOptBoolsDsc are set when called:
+//          b1      The first Basic Block with the BBJ_COND conditional jump type
+//          b2      The next basic block of b1 with the BBJ_RETURN
+//          b3      The next basic block of b2 with GT_RETURN
+//
+//      For example, (x==0 && y==0) generates:
+//          b1: GT_JTRUE (BBJ_COND)
+//          b2: GT_RETURN (BBJ_RETURN)
+//          b3: GT_RETURN (BBJ_RETURN),
+//      and it is folded into
+//          b1: GT_RETURN (BBJ_RETURN)
+//
+//      if b1.bbJumpDest == b2.bbNext, it transforms
+//          b1 : brtrue(t1, b3)
+//          b2 : ret(t2)
+//          b3 : ret(0)
+//      to
+//          b1 : ret((!t1) && t2)
+//
+bool Compiler::optOptimizeBoolsReturnBlock(OptBoolsDsc* pOptBoolsDsc)
+{
+    bool changeReturn = false;
+
+    BasicBlock* b1 = pOptBoolsDsc->b1;
+    BasicBlock* b2 = pOptBoolsDsc->b2;
+    BasicBlock* b3 = pOptBoolsDsc->b3;
+
+    pOptBoolsDsc->sameTarget = false;
+    Statement* const s1      = optOptimizeBoolsChkBlkCond(pOptBoolsDsc);
+    if (s1 == nullptr)
+    {
+        return changeReturn;
+    }
+
+    // Find the branch conditions of b1 and b2
+
+    GenTree* const c1 = optIsBoolComp(&pOptBoolsDsc->t1);
+    if (c1 == nullptr)
+    {
+        return changeReturn;
+    }
+    pOptBoolsDsc->c1 = c1;
+
+    GenTree* const c2 = optIsBoolComp(&pOptBoolsDsc->t2);
+    if (c2 == nullptr)
+    {
+        return changeReturn;
+    }
+    pOptBoolsDsc->c2 = c2;
+
+    // Find the type and cost conditions of t1 and t2
+
+    if (!optOptimizeBoolsChkTypeCostCond(pOptBoolsDsc))
+    {
+        return changeReturn;
+    }
+
+    // Get the fold operator and the comparison operator
+
+    var_types foldType = c1->TypeGet();
+    if (varTypeIsGC(foldType))
+    {
+        foldType = TYP_I_IMPL;
+    }
+    pOptBoolsDsc->foldType = foldType;
+
+    pOptBoolsDsc->foldOp = GT_NONE;
+    pOptBoolsDsc->cmpOp  = GT_NONE;
+
+    optReturnGetFoldAndCompOper(pOptBoolsDsc);
+    if (pOptBoolsDsc->foldOp == GT_NONE || pOptBoolsDsc->cmpOp == GT_NONE)
+    {
+        return changeReturn;
+    }
+
```

Comment:
Make `optReturnGetFoldAndCompOper` return `false` on "failure"? Actually, since there's only a single call site, is it really worth calling a function?

```suggestion
    if (!optReturnGetFoldAndCompOper(pOptBoolsDsc))
    {
        return false;
    }
```

Reply:
Done. Removed the function and moved it into `optOptimizeBoolsReturnBlock()`.